### PR TITLE
RPC

### PR DESCRIPTION
**Pull request by chasehensel**
@awans I think this is ready for a review!

### CODE REVIEW

File: client/catalog/src/app/rpc/RPCDetail.svelte [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+var name = "code";
+var code = "";
+
+load.then(obj => {
+	breadcrumbStore.set(
+		[{
+			href: "/rpcs",
+			text: "RPCs",
+		}, {
+			href: "/rcp/" + id,
```
Comment (awans) [inline client/catalog/src/app/rpc/RPCDetail.svelte:28]: typo?

File: internal/db/code.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ type FunctionSignature int64
 
 const (
 	FromJSON FunctionSignature = iota
+	RPC
```
Comment (awans) [inline internal/db/code.go:22]: Nice!

File: internal/db/db.go [hunk 3]
```diff
@@ -143,115 +144,253 @@ func (tx *holdTx) FindOne(modelID uuid.UUID, matcher Matcher) (rec Record, err e
...
-	return hits
+	tx.h = h
+	return nil
 }
 
-func (tx *holdTx) Insert(rec Record) {
+func (tx *holdTx) Update(r Record, fields map[string]interface{}) (Record, error) {
 	tx.ensureWrite()
-	tx.h = tx.h.Insert(rec)
+	//Need to load a model for the record
```
Comment (awans) [inline internal/db/db.go:164]: please gofmt this

File: internal/db/db.go [hunk 4]
```diff
@@ -143,115 +144,253 @@ func (tx *holdTx) FindOne(modelID uuid.UUID, matcher Matcher) (rec Record, err e
...
+	if err != nil {
+		return err
 	}
-	return hits
+	tx.h = h
+	return nil
 }
 
-func (tx *holdTx) Insert(rec Record) {
+func (tx *holdTx) Update(r Record, fields map[string]interface{}) (Record, error) {
```
Comment (awans) [inline internal/db/db.go:162]: I'm not sure if the interface at the DB level should be a map[string]interface{}.. that's leaking the JSON-ness of the API into the DB a bit. I wonder if the API should just be
`Update(id uuid.UUID, newRecord db.Record)`

File: internal/db/db.go [hunk 5]
```diff
@@ -143,115 +144,253 @@ func (tx *holdTx) FindOne(modelID uuid.UUID, matcher Matcher) (rec Record, err e
...
 	tx.ensureWrite()
-	tx.h = tx.h.Insert(rec)
+	//Need to load a model for the record
+	m, _ := tx.GetModel(r.Type())
+	newRec := r.DeepCopy(m)
```
Comment (awans) [inline internal/db/db.go:166]: curious why we have to pass the model into DeepCopy. I think that's worth fixing before we merge this.

File: internal/db/db.go [hunk 6]
```diff
@@ -143,115 +144,253 @@ func (tx *holdTx) FindOne(modelID uuid.UUID, matcher Matcher) (rec Record, err e
...
+			return err
+		}
+		err = left.SetFK(rel.LeftName, id)
+		if err != nil {
+			return err
+		}
 	} else if rel.RightBinding == BelongsTo && (rel.LeftBinding == HasOne || rel.LeftBinding == HasMany) {
 		// FK right
-		right.SetFK(rel.RightName, left.ID())
+		id, err := left.ID()
```
Comment (awans) [inline internal/db/db.go:196]: I wonder if it makes sense to panic on an error for ID. 

If a record doesn't have an ID field, something has gone truly awry.

File: internal/db/db.go [hunk 7]
```diff
@@ -143,115 +144,253 @@ func (tx *holdTx) FindOne(modelID uuid.UUID, matcher Matcher) (rec Record, err e
...
-	return Relationship{
-		ID:           storeRel.ID(),
-		LeftBinding:  RelType(storeRel.Get("leftBinding").(int64)),
-		LeftModelID:  storeRel.GetFK("leftModel"),
-		LeftName:     storeRel.Get("leftName").(string),
-		RightBinding: RelType(storeRel.Get("rightBinding").(int64)),
-		RightModelID: storeRel.GetFK("rightModel"),
-		RightName:    storeRel.Get("rightName").(string),
+func LoadRel(storeRel Record) (Relationship, error) {
+	id, err := storeRel.ID()
```
Comment (awans) [inline internal/db/db.go:223]: boy, this sure is noisy. But I think you can clean it up pretty nicely by following the `errWriter` pattern in this golang.org blog post: https://blog.golang.org/errors-are-values

(RecordReader or something like that)

File: internal/db/hold.go [hunk 8]
```diff
@@ -36,43 +36,52 @@ func (h *Hold) FindOne(modelID uuid.UUID, q Matcher) (Record, error) {
 	return nil, ErrNotFound
 }
 
-type MatchIter struct {
-	q  Matcher
-	it *iradix.Iterator
-}
-
-func (mi MatchIter) Next() (Record, bool) {
```
Comment (awans) [inline internal/db/hold.go:44]: Hm. I think it's fine to add the FindMany convenience method, but I'd like to keep the MatchIter method around. I think it'll be useful down the line.

File: internal/db/q.go [hunk 9]
```diff
@@ -25,7 +25,10 @@ type FieldMatcher struct {
 
 // could be faster probably
 func (fm FieldMatcher) Match(st Record) (bool, error) {
-	candidate := st.Get(fm.field)
+	candidate, err := st.Get(fm.field)
```
Comment (awans) [inline internal/db/q.go:28]: damn, checking err here is probably actually pretty expensive. Maybe we should give record a MustGet method that just panics on error.

File: internal/db/q.go [hunk 10]
```diff
@@ -41,7 +44,10 @@ type FKFieldMatcher struct {
 }
 
 func (fm FKFieldMatcher) Match(st Record) (bool, error) {
-	candidate := st.GetFK(fm.field)
+	candidate, err := st.GetFK(fm.field)
```
Comment (awans) [inline internal/db/q.go:47]: (ditto)

File: internal/db/record.go [hunk 11]
```diff
@@ -184,3 +206,84 @@ func RecordForModel(m Model) Record {
 
 	return &rRec{St: st, M: &m}
 }
+
+//Uses this gist https://gist.github.com/hvoecking/10772475
+func translate(obj interface{}) interface{} {
```
Comment (awans) [inline internal/db/record.go:211]: Understanding this now -- we don't need it, even if down the line we support struct fields. Let's get rid of it, and just copy the St

File: internal/oplog/logged_db.go [hunk 12]
```diff
@@ -138,39 +146,55 @@ func (tx *loggedTx) GetModel(modelName string) (db.Model, error) {
...
 
-func (tx *loggedTx) FindMany(modelID uuid.UUID, matcher db.Matcher) []db.Record {
+func (tx *loggedTx) FindMany(modelID uuid.UUID, matcher db.Matcher) ([]db.Record, error) {
 	return tx.inner.FindMany(modelID, matcher)
 }
 
 func (tx *loggedTx) Commit() (err error) {
 	err = tx.l.Log(tx.txe)
 	if err != nil {
+		panic(err)
```
Comment (awans) [inline internal/oplog/logged_db.go:197]: why panic?

File: internal/rpc/eval.go [hunk 13]
```diff
@@ -0,0 +1,27 @@
+package rpc
+
+import (
+	"awans.org/aft/internal/db"
+	"awans.org/aft/internal/starlark"
+)
+
+func eval(name string, args map[string]interface{}, tx db.RWTx) (interface{}, error) {
```
Comment (awans) [inline internal/rpc/eval.go:8]: pretty clean!

File: internal/rpc/models.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+	ID:           uuid.MustParse("9221119b-495a-449c-b2b3-2c6610f89d7b"),
+	LeftModelID:  uuid.MustParse("29209517-1c39-4be9-9808-e1ed8e40c566"), // rpc
+	LeftName:     "code",
+	LeftBinding:  db.BelongsTo,
+	RightModelID: uuid.MustParse("8deaec0c-f281-4583-baf7-89c3b3b051f3"), // code
+	RightName:    "rpc",
+	RightBinding: db.HasOne,
+}
+
+var reactFormRPC = db.Code{
```
Comment (awans) [inline internal/rpc/models.go:32]: obviously this should get pulled out :)

File: internal/starlark/db.go [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+//Wrapper for the Record interface so we can control which methods to expose.
+// This gets surfaced in Starlark as return values of database functions
+type Record interface {
+	ID() (uuid.UUID, error)
+	Get(string) (interface{}, error)
+	GetFK(string) (uuid.UUID, error)
+}
+
+type rRec struct {
```
Comment (awans) [inline internal/starlark/db.go:60]: maybe a different name for this one. starlarkRecord?

File: internal/db/db.go (hunk 4 above)
Comment (chasehensel) [reply internal/db/db.go:162]: Why do you need the id? the record should already have it

**Pull request merged**